### PR TITLE
Fix compiler warnings, part 8

### PR DESCRIPTION
**Pull request by nilason**
Fixes compiler warnings:

- -Wincompatible-pointer-types-discards-qualifiers
- -Wunused-result
- -Wheader-guard
- "too many arguments in call to function"

Eighth part addressing #1247.

Note: there are a few cases of OGR error handling added, please do pay special attention to them and their implementation.

Modules / code parts directly affected:

- lib/raster
- lib/imagery
- lib/nviz
- lib/vector/Vlib
- db/drivers/ogr
- general/g.parser
- raster/r.fill.stats
- raster3d/r3.retile
- raster/r.horizon
- raster/r.random
- vector/v.lrs/v.lrs.create
- vector/v.lrs/v.lrs.label
- vector/v.lrs/v.lrs.segment
- vector/v.lrs/v.lrs.where
- vector/v.rectify

### CODE REVIEW

File: db/drivers/ogr/execute.c [hunk 1]
```diff
@@ -109,7 +109,9 @@ int db__driver_execute_immediate(dbString * sql)
 	    }
 	    OGR_F_SetFieldString(hFeature, cols[i].index, value);
 	}
-	OGR_L_SetFeature(hLayer, hFeature);
+	if (0 != OGR_L_SetFeature(hLayer, hFeature))
```
Comment (metzm) [inline db/drivers/ogr/execute.c:112]:
```suggestion
	if (OGR_L_SetFeature(hLayer, hFeature) != OGRERR_NONE)
```

File: lib/nviz/draw.c [hunk 2]
```diff
@@ -230,13 +230,13 @@ int Nviz_draw_all(nv_data * data)
 	Nviz_draw_all_surf(data);
 
     if (draw_vect)
-	Nviz_draw_all_vect(data);
+	Nviz_draw_all_vect();
```
Comment (metzm) [inline lib/nviz/draw.c:233]: The fn definition `int Nviz_draw_all_vect()` in L121 should be changed to `int Nviz_draw_all_vect(void)`

File: lib/nviz/draw.c [hunk 3]
```diff
@@ -230,13 +230,13 @@ int Nviz_draw_all(nv_data * data)
...
 
     if (draw_site)
-	Nviz_draw_all_site(data);
+	Nviz_draw_all_site();
```
Comment (metzm) [inline lib/nviz/draw.c:236]: The fn definition `int Nviz_draw_all_site()` in L146 should be changed to `int Nviz_draw_all_site(void)`

File: lib/nviz/draw.c [hunk 4]
```diff
@@ -230,13 +230,13 @@ int Nviz_draw_all(nv_data * data)
...
 
     if (draw_vol)
-	Nviz_draw_all_vol(data);
+	Nviz_draw_all_vol();
```
Comment (metzm) [inline lib/nviz/draw.c:239]: The fn definition `int Nviz_draw_all_vol()` in L175 should be changed to `int Nviz_draw_all_vol(void)`

File: lib/raster/put_row.c [hunk 5]
```diff
@@ -514,7 +514,7 @@ static void write_null_bits_compressed(const unsigned char *flags,
     compressed_buf = G_malloc(cmax);
 
     /* compress null bits file with LZ4, see lib/gis/compress.h */
-    nwrite = G_compress(flags, size, compressed_buf, cmax, 3);
+    nwrite = G_compress((char *)flags, size, compressed_buf, cmax, 3);
```
Comment (metzm) [inline lib/raster/put_row.c:517]: This cast from `const unsigned char *` to `char *` fixes compiler warnings but masks the need to synchronize the affected GRASS function definitions. Something for GRASS 8.

File: lib/vector/Vlib/open_ogr.c [hunk 6]
```diff
@@ -100,8 +100,13 @@ int V1_open_old_ogr(struct Map_info *Map, int update)
     G_debug(2, "OGR layer %d opened", layer);
 
     ogr_info->layer = Ogr_layer;
-    if (update && OGR_L_TestCapability(ogr_info->layer, OLCTransactions))
-	OGR_L_StartTransaction(ogr_info->layer);
+    if (update && OGR_L_TestCapability(ogr_info->layer, OLCTransactions) &&
+	(0 != OGR_L_StartTransaction(ogr_info->layer))) {
```
Comment (metzm) [inline lib/vector/Vlib/open_ogr.c:104]:
```suggestion
	( OGR_L_StartTransaction(ogr_info->layer) != OGRERR_NONE)) {
```

File: lib/vector/Vlib/write_ogr.c [hunk 7]
```diff
@@ -347,8 +347,12 @@ int create_ogr_layer(struct Map_info *Map, int type)
 		      "Unable to write attributes."));
     }
     
-    if (OGR_L_TestCapability(ogr_info->layer, OLCTransactions))
-	OGR_L_StartTransaction(ogr_info->layer);
+    if (OGR_L_TestCapability(ogr_info->layer, OLCTransactions) &&
+	(0 != OGR_L_StartTransaction(ogr_info->layer))) {
```
Comment (metzm) [inline lib/vector/Vlib/write_ogr.c:351]:
```suggestion
	(OGR_L_StartTransaction(ogr_info->layer) != OGRERR_NONE)) {
```

File: raster/r.horizon/main.c [hunk 8]
```diff
@@ -393,7 +393,7 @@ int main(int argc, char *argv[])
 	sscanf(parm.direction->answer, "%lf", &single_direction);
 
 
-    if (isMode(WHOLE_RASTER)) {
+    if (WHOLE_RASTER == isMode()) {
```
Comment (metzm) [inline raster/r.horizon/main.c:396]: `int  isMode()` in L132 should be changed to `int  isMode(void)`

File: vector/v.lrs/v.lrs.create/main.c [hunk 9]
```diff
@@ -206,14 +206,14 @@ int main(int argc, char **argv)
     driver_opt->required = NO;
     driver_opt->description = _("Driver name for reference system table");
     driver_opt->options = db_list_drivers();
-    driver_opt->answer = db_get_default_driver_name();
+    driver_opt->answer = (char *) db_get_default_driver_name();
```
Comment (metzm) [inline vector/v.lrs/v.lrs.create/main.c:209]: Maybe `option->answer` in [include/gis.h](https://github.com/OSGeo/grass/blob/master/include/gis.h#L541) can be changed to `const char *`? Together with `option->answers` in  [include/gis.h](https://github.com/OSGeo/grass/blob/master/include/gis.h#L543). Something for GRASS 8.

IMHO, these type casts avoid compiler warnings but do not fix the real problem.

File: db/drivers/ogr/execute.c (hunk 1 above)
Comment (nilason) [reply db/drivers/ogr/execute.c]: Good. I actually did consider using `OGRERR_NONE`,  but sided with using example from surrounding code.
Updated.

File: lib/nviz/draw.c (hunk 2 above)
Comment (nilason) [reply lib/nviz/draw.c:233]: updated

File: lib/nviz/draw.c (hunk 3 above)
Comment (nilason) [reply lib/nviz/draw.c:236]: updated

File: lib/nviz/draw.c (hunk 4 above)
Comment (nilason) [reply lib/nviz/draw.c:239]: updated

File: lib/raster/put_row.c (hunk 5 above)
Comment (nilason) [reply lib/raster/put_row.c]: That should have been `const unsigned char *` to `unsigned char *` cast, `const` being the "only" difference. Updated accordingly.

I guess the G_compress() family of functions could be updated from:
```
int G_compress(unsigned char *, int, unsigned char *, int, int);
```

to
```
int G_compress(const unsigned char *, int, unsigned char *, int, int);
```
where the source is a const pointer. Is this what you were thinking of?

File: lib/vector/Vlib/open_ogr.c (hunk 6 above)
Comment (nilason) [reply lib/vector/Vlib/open_ogr.c]: updated

File: lib/vector/Vlib/write_ogr.c (hunk 7 above)
Comment (nilason) [reply lib/vector/Vlib/write_ogr.c]: updated

File: raster/r.horizon/main.c (hunk 8 above)
Comment (nilason) [reply raster/r.horizon/main.c:396]: updated

File: vector/v.lrs/v.lrs.create/main.c (hunk 9 above)
Comment (nilason) [reply vector/v.lrs/v.lrs.create/main.c:209]: I don't know the reason Option->answer is not a const string. But I'm afraid changing that wouldn't be a stroll in the park :)  These casts are all over. A alternative now, for this `db_get_default_driver_name()` case, could be to use G_store() (with the memory leak that comes with it).

Comment (nilason) [reply vector/v.lrs/v.lrs.create/main.c:209]: Just out of curiosity I tried compile with Option->answer defined as `const char *answer`.
This resulted in some 660 warnings of type `-Wincompatible-pointer-types-discards-qualifiers`.

File: db/drivers/ogr/execute.c (hunk 1 above)
Comment (metzm) [reply db/drivers/ogr/execute.c]: I recommend following the OGR API for [OGR_L_SetFeature()](https://gdal.org/doxygen/ogr__api_8h.html#a5b380520d0999422987f06b6e31b9b00)

File: lib/nviz/draw.c (hunk 2 above)
Comment (metzm) [reply lib/nviz/draw.c:233]: What about the suggested change?

File: lib/nviz/draw.c (hunk 3 above)
Comment (metzm) [reply lib/nviz/draw.c:236]: What about the suggested change?

File: lib/nviz/draw.c (hunk 4 above)
Comment (metzm) [reply lib/nviz/draw.c:239]: What about the suggested change?

File: lib/raster/put_row.c (hunk 5 above)
Comment (metzm) [reply lib/raster/put_row.c]: Some fns of external libraries used by the `G_compress()` family of fns want `(char *)` as input, e.g. `BZ2_bzBuffToBuffCompress()`. According casts are done in the respective `lib/gis/cmpr*.c` files.

Your updated change makes sense.

File: lib/nviz/draw.c (hunk 2 above)
Comment (metzm) [reply lib/nviz/draw.c:233]: Apparently resolved.

File: lib/nviz/draw.c (hunk 4 above)
Comment (metzm) [reply lib/nviz/draw.c:239]: Apparently resolved.

File: lib/nviz/draw.c (hunk 3 above)
Comment (metzm) [reply lib/nviz/draw.c:236]: Apparently resolved.

File: vector/v.lrs/v.lrs.create/main.c (hunk 9 above)
Comment (metzm) [reply vector/v.lrs/v.lrs.create/main.c:209]: OK for your change, but still, type casts avoid compiler warnings but do not fix the real problem.

**Comment by nilason** [comment]
Updated commit messages and I believe it is ready for merge.

I did split the original "-Wincompatible-pointer-types-discards-qualifiers" fix commit into two, depending on the cause of the warning (and nature of fix).

A note regarding the cast of the result of `db_get_default_driver_name()`. The function is defined as: 
```c
const char *db_get_default_driver_name(void)
```
but in fact returns a `char *`:

https://github.com/OSGeo/grass/blob/51fe070ae52ba63ab789bf30adf5808ce80dd5e9/lib/db/dbmi_base/default_name.c#L32

which is here then re-casted back to `char *`.

**Comment by metzm** [comment]
> A note regarding the cast of the result of `db_get_default_driver_name()`. The function is defined as:
> 
> ```c
> const char *db_get_default_driver_name(void)
> ```
> 
> but in fact returns a `char *`:
> 
> https://github.com/OSGeo/grass/blob/51fe070ae52ba63ab789bf30adf5808ce80dd5e9/lib/db/dbmi_base/default_name.c#L32
> 
> which is here then re-casted back to `char *`.

IMHO, `db_get_default_driver_name()` is correct, it provides the default driver name which should not be modified, thus `const`. Within the scope of this PR, I think re-casting back to `char *` is fine.

**Review by metzm — APPROVED** [review]
Ready to merge, thanks a lot!

**Comment by nilason** [comment]
> IMHO, db_get_default_driver_name() is correct, it provides the default driver name which should not be modified, thus const. Within the scope of this PR, I think re-casting back to char * is fine.

Oh, I do understand the logic that db_get_default_driver_name() is a const and agree it's correct. Only brought it up because I suggested the return value could be wrapped in a G_store() instead of casting, but that wouldn't make much sense in light of its implementation.

**Comment by nilason** [comment]
Cherry picked and pushed to master.
Thank you!

**Comment by nilason** [comment]
@metzm 
Regarding the structure member Option->answer and the casting problems it seem to cause. I noticed that originally (at least with G5) all (string) members were `char *`, but this [changed in 2008](https://trac.osgeo.org/grass/changeset/31906) (G6.4) to what we see today. Only `answer` (and `answers`) stayed mutable. The question is: do you think it's worth the effort to even explore the possibility to change ´answer´ to a const? Or will that set me off in a one-way road?

**Pull request closed without merging**